src/engine/Union.cpp

```diff
@@ -77,10 +77,15 @@ Union::Union(QueryExecutionContext* qec,
 
 string Union::getCacheKeyImpl() const {
   std::ostringstream os;
+  os << "{\n";
   os << _subtrees[0]->getCacheKey() << "\n";
-  os << "UNION\n";
+  os << "} UNION {\n";
   os << _subtrees[1]->getCacheKey() << "\n";
-  os << "sort order: ";
+  os << "} column origins: ";
+  for (auto [left, right] : _columnOrigins) {
+    os << '(' << left << ", " << right << ") ";
+  }
+  os << " sort order: ";
```

Comment:
Can you briefly explain why this is necessary?

In particular: Does not the cache key for each side also provide the information which columns are visible, in case not all columns are visible?

Reply:
Good question, no it doesn't. Hidden variables generally don't have any effect the cache key. You can easily verify this by executing a subquery that selects a certain variable, and then running the same thing again but with `*` as a wildcard. You will see that the result will be retrieved from the cache and not get recomputed (all variables were already computed the first time, they just were not exported).
But for UNION it does affect the amount of rows the query produces, so it is important here. Other operations like Join have to do this too for their join column.